File: python/twelve-days/twelve_days.py
# -*- coding: utf-8 -*-

ORDINALS = ['zeroth', 'first', 'second', 'third', 'fourth', 'fifth', 'sixth',
            'seventh', 'eighth', 'ninth', 'tenth', 'eleventh', 'twelfth']
RECEPTION = 'On the {nth} day of Christmas my true love gave to me: {presents}.'
GIFTS = [
    None,
    'a Partridge in a Pear Tree', 'two Turtle Doves',
    'three French Hens', 'four Calling Birds',
    'five Gold Rings', 'six Geese-a-Laying',
    'seven Swans-a-Swimming', 'eight Maids-a-Milking',
    'nine Ladies Dancing', 'ten Lords-a-Leaping',
    'eleven Pipers Piping', 'twelve Drummers Drumming'
]

def recite(start_verse: int, end_verse: int) -> list:
    """Recite the <opinion>annoying</opinion> twelve days of christmas carol.
    """

    if start_verse != end_verse: # follow the tests' example and recurse
        return [recite(n, n)[0] for n in range(start_verse, end_verse + 1)]

    return [RECEPTION.format(nth=ORDINALS[start_verse],
                             presents=_presents_phrase(GIFTS[start_verse:0:-1]))]

def _presents_phrase(items: list) -> str:
    "Return a presentable string on the presents for nth day"
    *items, last_item = items
    if not items:
        return last_item
    return ', '.join(items) + ', and ' + last_item

File: ruby/twelve-days/twelve_days.rb
# frozen_string_literal: true

class TwelveDays
  ORDINALS = %w[first second third fourth fifth sixth
                seventh eighth ninth tenth eleventh twelfth].freeze
  RECEPTION =
    'On the %<nth_day>s day of Christmas my true love gave to me: %<presents>s.'

  GIFTS = [
    'a Partridge in a Pear Tree', 'two Turtle Doves', 'three French Hens',
    'four Calling Birds', 'five Gold Rings', 'six Geese-a-Laying',
    'seven Swans-a-Swimming', 'eight Maids-a-Milking', 'nine Ladies Dancing',
    'ten Lords-a-Leaping', 'eleven Pipers Piping', 'twelve Drummers Drumming'
  ].freeze

  def self.song(upto_day = 12)
    new(upto_day).song
  end

  def initialize(upto_day)
    @upto_day = upto_day
  end

  def song
    @upto_day.times.map { |day| verse(day) }.join("\n\n") + "\n"
  end

  def song_simpler
    IO.read(File.expand_path('song.txt', __dir__))
  end

  private

  #
  # Recite a single line about what your true love gave you on the nth day.
  #
  def verse(day)
    format(RECEPTION, nth_day: ORDINALS[day], presents: to_phrase(gifts(day)))
  end

  #
  # Return a presentable string on the presents for nth day
  #
  def to_phrase(items)
    *items, last_item = items
    return last_item if items.empty?

    items.join(', ') + ', and ' + last_item
  end

  def gifts(day)
    GIFTS.take(day + 1).reverse
  end
end
